File: homework4_1/src/main/java/by/it/academy/homework4_1/MainActivity.java
package by.it.academy.homework4_1;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import android.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;
import android.view.View;
import android.view.inputmethod.EditorInfo;

import java.util.ArrayList;


public class MainActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private Contact contact;
    private ContactAdapter contactAdapter;
    private static final int ADD_CONTACT = 0;
    private static final int EDIT_CONTACT = 1;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbarMain);
        setSupportActionBar(toolbar);
        recyclerView = findViewById(R.id.recyclerView);
        SearchView searchView = findViewById(R.id.searchViewActivityMain);

        if (savedInstanceState != null) {
            contactAdapter = new ContactAdapter(savedInstanceState.getParcelableArrayList("list"));
        } else {
            contactAdapter = new ContactAdapter(new ArrayList<>());
        }


        recyclerView.setAdapter(contactAdapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(this, RecyclerView.VERTICAL, false));
        checkContacts(contactAdapter);
        contactAdapter.setListener(new ContactAdapter.Listener() {

            @Override
            public void onClick(int position) {
                Intent intent = new Intent(MainActivity.this, EditContact.class);
                contact = contactAdapter.getContactList().get(position);
                setResult(Activity.RESULT_OK, intent);
                intent.putExtra("contact", contact);
                intent.putExtra("position", position);
                startActivityForResult(intent, EDIT_CONTACT);
            }
        });

        searchView.setImeOptions(EditorInfo.IME_ACTION_DONE);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                contactAdapter.getFilter().filter(newText);
                return false;
            }
        });


        findViewById(R.id.floatingActionButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(MainActivity.this, AddContact.class);
                startActivityForResult(intent,ADD_CONTACT);
            }
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        Contact nContact;

        if (resultCode == RESULT_OK && requestCode == ADD_CONTACT) {
            assert data != null;
            nContact = data.getParcelableExtra("add_contact");
            contactAdapter.add(nContact);
            checkContacts(contactAdapter);


        } else if (resultCode == RESULT_OK && requestCode == EDIT_CONTACT) {
            assert data != null;
            nContact = data.getParcelableExtra("edit_contact");
            int position = data.getIntExtra("position", 0);

            if (nContact != null) {
                contactAdapter.edit(position, nContact);
                checkContacts(contactAdapter);
            } else {
                recyclerView.invalidate();
                checkContacts(contactAdapter);
                contactAdapter.remove(data.getIntExtra("position", 0));
            }
        }
    }

    public void checkContacts(@NonNull ContactAdapter contactAdapter) {
        if (contactAdapter.getContactList().isEmpty()) {
            findViewById(R.id.ifNoContacts).setVisibility(View.VISIBLE);
        } else {
            findViewById(R.id.ifNoContacts).setVisibility(View.INVISIBLE);
        }
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        if (contactAdapter.getContactList().size() != 0) {
            outState.putParcelableArrayList("list", (ArrayList<? extends Parcelable>) contactAdapter.getContactList());
        } else {
            outState.putParcelableArrayList("list", new ArrayList<>());
        }
    }
}

File: homework-7_rx_java/src/main/java/by/it/academy/homework_7_rx_java/activities/MainActivity.kt
package by.it.academy.homework_7_rx_java.activities

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.Menu
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.appcompat.widget.SearchView
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import by.it.academy.homework_7_rx_java.CarInfoAdapter
import by.it.academy.homework_7_rx_java.R
import by.it.academy.homework_7_rx_java.database.CarInfoDAO
import by.it.academy.homework_7_rx_java.database.DataBaseCarInfo
import com.google.android.material.floatingactionbutton.FloatingActionButton
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date

private const val ADD_ACTIVITY_CODE = 1
private const val EDIT_ACTIVITY_CODE = 2
private const val WORK_LIST_ACTIVITY = 10
private const val RESULT_CODE_BUTTON_BACK = 5

class MainActivity : AppCompatActivity() {
    private lateinit var fab: FloatingActionButton
    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: CarInfoAdapter
    private lateinit var noCarsAddedText: TextView
    private lateinit var database: DataBaseCarInfo
    private lateinit var carInfoDAO: CarInfoDAO
    private lateinit var searchView: SearchView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_car_list)
        fab = findViewById(R.id.addCarFab)
        recyclerView = findViewById(R.id.carListRecycler)
        noCarsAddedText = findViewById(R.id.emptyCarList_text)
        initDatabase()
        setRecyclerSettings()
        setFabListener()
        setAdapterListeners()
        writeLogToFile()
        setNoCarsTextViewVisibility()
    }

    private fun initDatabase() {
        database = DataBaseCarInfo.getDataBase(this)
        carInfoDAO = database.getCarInfoDAO()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.main_menu, menu)
        searchView = menu?.findItem(R.id.search)?.actionView as SearchView
        searchView.apply {
            imeOptions = EditorInfo.IME_ACTION_DONE
            setOnQueryTextListener(object : SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(p0: String?) = false

                override fun onQueryTextChange(p0: String?): Boolean {
                    adapter.filter.filter(p0)
                    return false
                }
            })
        }
        return true
    }

    private fun setFabListener() {
        fab.setOnClickListener {
            val intent = Intent(this, ActivityAddCar::class.java)
            startActivityForResult(intent, ADD_ACTIVITY_CODE)
        }
    }

    private fun setRecyclerSettings() {
        val allDataList = carInfoDAO.getAll().sortedBy { it.producer.toLowerCase() }
        adapter = CarInfoAdapter(allDataList)
        recyclerView.adapter = adapter
        recyclerView.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode != RESULT_CODE_BUTTON_BACK) {
            adapter.updateList(carInfoDAO.getAll().sortedBy { it.producer.toLowerCase() })
            if (!searchView.isIconified) {
                searchView.onActionViewCollapsed()
            }
        }
        setNoCarsTextViewVisibility()
    }

    private fun writeLogToFile() {
        val logList = StringBuilder()
        val file = File(filesDir, "Logs.txt")
        if (!file.exists()) {
            file.createNewFile()
        }
        logList.append("${file.readText()} \n")
        val simpleDateFormat = SimpleDateFormat.getDateTimeInstance()
        val date = simpleDateFormat.format(Date())
        logList.append(date)
        file.writeText(logList.toString())
    }

    private fun setNoCarsTextViewVisibility() {
        if (adapter.itemCount != 0) noCarsAddedText.visibility = View.INVISIBLE
        else noCarsAddedText.visibility = View.VISIBLE
    }

    private fun setAdapterListeners() {
        adapter.onEditIconClickListener = {
            val intent = Intent(this, ActivityEditCar::class.java)
            intent.putExtra("carInfo", it)
            startActivityForResult(intent, EDIT_ACTIVITY_CODE)
        }
        adapter.onCarInfoShowWorkListClickListener = {
            val intent = Intent(this, ActivityWorkList::class.java)
            intent.putExtra("carInfo", it)
            startActivityForResult(intent, WORK_LIST_ACTIVITY)
        }
    }
}